internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,256 @@
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"maps"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Watcher interface {
+	Watch(source.Source) error
+}
+
+// Cache is a storage mechanism for keeping track of
+// managed content sources. It also exposes methods
+// for retrieving the stored references of the managed
+// content
+type Cache interface {
+	io.Closer
+	// Watch establishes watches for all provided client.Objects
+	Watch(context.Context, Watcher, ...client.Object) error
+}
+
+// CloserSyncingSource is a wrapper of the controller-runtime
+// source.SyncingSource that includes methods for:
+//   - Getting a specific resource synced by the source
+//   - Listing resources synced by the source
+//   - Closing the source, stopping it's interaction with the Kubernetes API server and reaction to events
+type CloserSyncingSource interface {
+	source.SyncingSource
+	io.Closer
+}
+
+type Sourcerer interface {
+	Source(schema.GroupVersionKind, client.Object, func(context.Context)) (CloserSyncingSource, error)
+}
+
+type cache struct {
+	sources     map[schema.GroupVersionKind]CloserSyncingSource
+	sourcerer   Sourcerer
+	owner       client.Object
+	syncTimeout time.Duration
+	mu          sync.Mutex
+}
+
+func NewCache(sourcerer Sourcerer, owner client.Object, syncTimeout time.Duration) Cache {
+	return &cache{
+		sources:     make(map[schema.GroupVersionKind]CloserSyncingSource),
+		sourcerer:   sourcerer,
+		owner:       owner,
+		syncTimeout: syncTimeout,
+	}
+}
+
+var _ Cache = (*cache)(nil)
+
+func (c *cache) Watch(ctx context.Context, watcher Watcher, objs ...client.Object) error {
+	if watcher == nil {
```

Comment:
can probably get rid of this check. if we pass in a nil watcher, we'll likely eventually panic when we try to use it, and i think that's an appropriate way to handle a programmer error.

Reply:
addressed in https://github.com/operator-framework/operator-controller/pull/1119/commits/af075c5137065a5e0b2e0cfc1b4debbb367f792f